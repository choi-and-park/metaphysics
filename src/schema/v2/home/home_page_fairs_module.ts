import Fair from "schema/v2/fair"
import { GraphQLList, GraphQLObjectType, GraphQLNonNull } from "graphql"
import { ResolverContext } from "types/graphql"
import { filter } from "lodash"
import moment from "moment"

export const HomePageFairsModuleType = new GraphQLObjectType<
  any,
  ResolverContext
>({
  name: "HomePageFairsModule",
  fields: {
    results: {
      type: new GraphQLNonNull(new GraphQLList(Fair.type)),
      resolve: (_root, _options, { fairsLoader }) => {
        // Check for all fairs that are currently running
        const gravityOptions = {
          has_full_feature: true,
          sort: "-start_at",
          active: true,
        }

        const now = moment.utc()

        return fairsLoader(gravityOptions).then(
          ({ body: unfilteredRunningFairs }) => {
            // Gravity returns fairs that are both current and upcoming.
            // Make sure only the current ones appear in the results list.
            const runningFairs = filter(unfilteredRunningFairs, (fair) => {
              const startAt = moment.utc(fair.start_at)
              return now.isAfter(startAt)
            })

            // If there are less than 8, get the most recent closed fairs
            if (runningFairs.length >= 8) {
              return runningFairs
            }

            const newOptions = {
              ...gravityOptions,
              status: "closed",
              active: false,
              size: 8 - runningFairs.length,
            }

            return fairsLoader(newOptions).then(
              ({ body: unfilteredClosedFairs }) => {
                const closedFairs = filter(unfilteredClosedFairs, (fair) => {
                  const endAt = moment.utc(fair.end_at)
                  return now.isAfter(endAt)
                })
                return runningFairs.concat(closedFairs)
              }
            )
          }
        )
      },
    },
  },
})

const HomePageFairsModule = {
  type: HomePageFairsModuleType,
  resolve: (_root, obj) => obj,
}

export default HomePageFairsModule
