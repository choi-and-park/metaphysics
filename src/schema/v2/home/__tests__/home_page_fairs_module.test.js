import { runQuery } from "schema/v2/test/utils"
import moment from "moment"
import { range } from "lodash"

// adding this dep explicitly so the tests run when we change the main file
;() => {
  require("../home_page_fairs_module")
}

let _counter = 0
const uniqid = () => _counter++

const mockRunningFair = () => {
  const id = uniqid()
  return {
    id: `running-fair-${id}`,
    default_profile_id: `running-fair-${id}`,
    start_at: moment().subtract(1, "day"),
    end_at: moment().add(1, "day"),
    name: `A running fair ${id}`,
    published: true,
    subtype: null,
    summary: "",
    layout: null,
    display_vip: false,
    has_full_feature: true,
  }
}

const mockFutureFair = () => {
  const id = uniqid()
  return {
    id: `future-fair-${id}`,
    default_profile_id: `future-fair-${id}`,
    start_at: moment().add(1, "day"),
    end_at: moment().add(10, "day"),
    name: `A future fair ${id}`,
    published: true,
    subtype: null,
    summary: "",
    layout: null,
    display_vip: false,
    has_full_feature: true,
  }
}

const mockPastFair = () => {
  const id = uniqid()
  return {
    id: `past-fair-${id}`,
    default_profile_id: `past-fair-${id}`,
    start_at: moment().subtract(10, "day"),
    end_at: moment().subtract(1, "day"),
    name: `A past fair ${id}`,
    published: true,
    subtype: null,
    summary: "",
    layout: null,
    display_vip: false,
    has_full_feature: true,
  }
}

const isRunningFair = (fair) => fair.name.startsWith("A running fair")
const isNotFutureFair = (fair) => !fair.name.startsWith("A future fair")

const runFairsQuery = async (query, config) =>
  runQuery(query, {
    fairsLoader: (options) => Promise.resolve(config(options)),
  })

describe("HomePageFairsModule", () => {
  it("shows the correct number of fairs", async () => {
    const runningFairs = [mockRunningFair()]

    const pastFairs = [mockPastFair()]

    const query = `
      {
        homePage {
          fairsModule {
            results {
              slug
              name
              isActive
            }
          }
        }
      }
    `

    const fairsModule = await runFairsQuery(query, (options) => ({
      body: options.active ? runningFairs : pastFairs,
    }))
    const results = fairsModule.homePage.fairsModule.results
    expect(results).toHaveLength(2)
  })

  it("does not request past fairs if it has 8 running ones", async () => {
    const runningFairs = range(8).map(() => mockRunningFair())

    const pastFairs = [mockPastFair()]

    const query = `
      {
        homePage {
          fairsModule {
            results {
              slug
              name
              isActive
            }
          }
        }
      }
    `

    const fairsModule = await runFairsQuery(query, (options) => ({
      body: options.active ? runningFairs : pastFairs,
    }))
    const results = fairsModule.homePage.fairsModule.results
    expect(results).toHaveLength(8)
    expect(results).not.toIncludeAnyMembers(pastFairs)
    expect(results).toSatisfyAll(isRunningFair)
  })

  it("doesn't return fairs that haven't opened yet", async () => {
    const query = `
      {
        homePage {
          fairsModule {
            results {
              name
            }
          }
        }
      }
	`

    const fairs = [mockRunningFair(), mockFutureFair()]

    const fairsModule = await runFairsQuery(query, () => ({ body: fairs }))
    const results = fairsModule.homePage.fairsModule.results
    expect(results).toHaveLength(1)
    expect(results).toSatisfyAll(isNotFutureFair)
  })
})
