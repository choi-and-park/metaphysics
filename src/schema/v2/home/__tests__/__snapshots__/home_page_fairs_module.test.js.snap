// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePageFairsModule does not request past fairs if it has 8 running ones 1`] = `
Object {
  "homePage": Object {
    "fairsModule": Object {
      "results": Array [
        Object {
          "isActive": false,
          "name": "Artissima 2017",
          "slug": "artissima-2017",
        },
        Object {
          "isActive": false,
          "name": "Artissima 2017",
          "slug": "artissima-2017",
        },
        Object {
          "isActive": false,
          "name": "Artissima 2017",
          "slug": "artissima-2017",
        },
        Object {
          "isActive": false,
          "name": "Artissima 2017",
          "slug": "artissima-2017",
        },
        Object {
          "isActive": false,
          "name": "Artissima 2017",
          "slug": "artissima-2017",
        },
        Object {
          "isActive": false,
          "name": "Artissima 2017",
          "slug": "artissima-2017",
        },
        Object {
          "isActive": false,
          "name": "Artissima 2017",
          "slug": "artissima-2017",
        },
        Object {
          "isActive": false,
          "name": "Artissima 2017",
          "slug": "artissima-2017",
        },
      ],
    },
  },
}
`;

exports[`HomePageFairsModule works 1`] = `
Object {
  "homePage": Object {
    "fairsModule": Object {
      "results": Array [
        Object {
          "isActive": false,
          "name": "Artissima 2017",
          "slug": "artissima-2017",
        },
        Object {
          "isActive": false,
          "name": "Zâ“ˆONAMACO FOTO & SALÃ“N DEL ANTICUARIO 2017",
          "slug": "zonamaco-foto-and-sal-n-del-anticuario-2017",
        },
      ],
    },
  },
}
`;
